File: src/main/java/gl3/GL_injection.java
package gl3;

import com.jogamp.newt.event.*;
import com.jogamp.newt.opengl.GLWindow;
import com.jogamp.opengl.*;
import com.jogamp.opengl.util.Animator;
import com.jogamp.opengl.util.GLBuffers;

import java.nio.FloatBuffer;

import static com.jogamp.opengl.GL2ES3.GL_COLOR;

/**
 * Created by GBarbieri on 27.03.2017.
 */
public class GL_injection implements GLEventListener, KeyListener {

    private static GLWindow window;
    private static Animator animator;

    public static void main(String[] args) {
        new GL_injection().setup();
    }

    private FloatBuffer clearColor = GLBuffers.newDirectFloatBuffer(new float[]{0, 0, 0, 0});

    private void setup() {

        GLProfile glProfile = GLProfile.get(GLProfile.GL3);
        GLCapabilities glCapabilities = new GLCapabilities(glProfile);

        window = GLWindow.create(glCapabilities);

        window.setTitle("gl injection");
        window.setSize(1024, 768);

        window.addGLEventListener(this);
        window.addKeyListener(this);

        window.setAutoSwapBufferMode(false);

        window.setVisible(true);

        animator = new Animator(window);
        animator.start();

        window.addWindowListener(new WindowAdapter() {
            @Override
            public void windowDestroyed(WindowEvent e) {
                animator.stop();
                System.exit(1);
            }
        });
    }

    @Override
    public void init(GLAutoDrawable drawable) {
    }

    @Override
    public void reshape(GLAutoDrawable drawable, int x, int y, int width, int height) {
    }

    @Override
    public void display(GLAutoDrawable drawable) {
    }
    
    @Override
    public void dispose(GLAutoDrawable drawable) {
    }

    @Override
    public void keyPressed(KeyEvent e) {

        if (e.getKeyCode() == KeyEvent.VK_ESCAPE)
            new Thread(() -> {
                window.destroy();
            }).start();

        switch (e.getKeyCode()) {
            case KeyEvent.VK_R:
                modified(0);
                break;
            case KeyEvent.VK_G:
                modified(1);
                break;
            case KeyEvent.VK_B:
                modified(2);
                break;
            case KeyEvent.VK_A:
                modified(3);
                break;
        }
    }

    private void modified(int index) {
        clearColor.put(index, (clearColor.get(index) == 0) ? 1 : 0);
        System.out.println("clear color: (" + clearColor.get(0) + ", " + clearColor.get(1) + ", " + clearColor.get(2) + ", " + clearColor.get(3) + ")");
        window.invoke(false, new GLRunnable() {
            @Override
            public boolean run(GLAutoDrawable drawable) {
                drawable.getGL().getGL3().glClearBufferfv(GL_COLOR, 0, clearColor);
                drawable.swapBuffers();
                return true;
            }
        });
    }


    @Override
    public void keyReleased(KeyEvent e) {
    }
}

File: src/main/kotlin/gl3/gl_injection.kt
package gl3

import com.jogamp.newt.event.KeyEvent
import com.jogamp.newt.event.KeyListener
import com.jogamp.newt.event.WindowAdapter
import com.jogamp.newt.event.WindowEvent
import com.jogamp.newt.opengl.GLWindow
import com.jogamp.opengl.*
import com.jogamp.opengl.GL2ES3.GL_COLOR
import com.jogamp.opengl.util.Animator
import com.jogamp.opengl.util.GLBuffers
import glm.set
import uno.buffer.toFloatBuffer

/**
 * Created by GBarbieri on 27.03.2017.
 */

fun main(args: Array<String>) {
    GL_injection_().setup()
}

class GL_injection_ : GLEventListener, KeyListener {

    lateinit var window: GLWindow
    lateinit var animator:Animator
    
    val clearColor = floatArrayOf(0f, 0f, 0f, 0f).toFloatBuffer()

    fun setup() {

        val glProfile = GLProfile.get(GLProfile.GL3)
        val glCapabilities = GLCapabilities(glProfile)

        window = GLWindow.create(glCapabilities)

        window.title = "gl injection"
        window.setSize(1024, 768)

        window.addGLEventListener(this)
        window.addKeyListener(this)

        window.autoSwapBufferMode = false

        window.isVisible = true

        animator = Animator(window)
        animator.start()

        window.addWindowListener(object : WindowAdapter() {
            override fun windowDestroyed(e: WindowEvent?) {
                animator.stop()
                System.exit(1)
            }
        })
    }

    override fun init(drawable: GLAutoDrawable) {}

    override fun reshape(drawable: GLAutoDrawable, x: Int, y: Int, width: Int, height: Int) {}

    override fun display(drawable: GLAutoDrawable) {}

    override fun dispose(drawable: GLAutoDrawable) {}

    override fun keyPressed(e: KeyEvent) {

        if (e.keyCode == KeyEvent.VK_ESCAPE)
            Thread { window.destroy() }.start()

        when (e.keyCode) {
            KeyEvent.VK_R -> modified(0)
            KeyEvent.VK_G -> modified(1)
            KeyEvent.VK_B -> modified(2)
            KeyEvent.VK_A -> modified(3)
        }
    }

    fun modified(index: Int) {
        clearColor[index] = if (clearColor[index] == 0f) 1f else 0f
        println("clear color: (" + clearColor[0] + ", " + clearColor[1] + ", " + clearColor[2] + ", " + clearColor[3] + ")")
        window.invoke(false) { drawable ->
            drawable.gl.gL3.glClearBufferfv(GL_COLOR, 0, clearColor)
            drawable.swapBuffers()
            true
        }
    }
    
    override fun keyReleased(e: KeyEvent) {}
}